### PR TITLE
Added `WorkloadIdentityCredential`

### PR DESCRIPTION
**Pull request by xiangyan99**

### CODE REVIEW

File: sdk/identity/azure-identity/azure/identity/aio/_credentials/default.py [hunk 1]
```diff
@@ -109,6 +113,13 @@ def __init__(self, **kwargs: Any) -> None:
         credentials = []  # type: List[AsyncTokenCredential]
         if not exclude_environment_credential:
             credentials.append(EnvironmentCredential(authority=authority, **kwargs))
+        if all(os.environ.get(var) for var in EnvironmentVariables.TOKEN_EXCHANGE_VARS):
+            client_id = managed_identity_client_id or os.environ.get(EnvironmentVariables.AZURE_CLIENT_ID)
```
Comment (chlowell) [inline sdk/identity/azure-identity/azure/identity/aio/_credentials/default.py:117]: The argument for this, if any, should be `workload_identity_client_id`.

File: sdk/identity/azure-identity/azure/identity/_credentials/workload_identity.py [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+    def __init__(
+            self,
+            tenant_id: str,
+            client_id: str,
+            file: str,
+            **kwargs: Any
+    ) -> None:
+        kwargs.pop("token_file_path", None)
+        self._credential = TokenExchangeCredential(
```
Comment (chlowell) [inline sdk/identity/azure-identity/azure/identity/_credentials/workload_identity.py:31]: Would be simpler to replace `TokenExchangeCredential` instead? Keeping it seems to add an unnecessary layer to this credential.

File: sdk/identity/azure-identity/azure/identity/aio/_credentials/default.py (hunk 1 above)
Comment (xiangyan99) [reply sdk/identity/azure-identity/azure/identity/aio/_credentials/default.py:117]: managed_identity_client_id will be used in the managed identity route, right?

Shouldn't we keep the same behavior to honor it?

File: sdk/identity/azure-identity/azure/identity/aio/_credentials/default.py [hunk 3]
```diff
@@ -109,6 +113,13 @@ def __init__(self, **kwargs: Any) -> None:
         credentials = []  # type: List[AsyncTokenCredential]
         if not exclude_environment_credential:
             credentials.append(EnvironmentCredential(authority=authority, **kwargs))
+        if all(os.environ.get(var) for var in EnvironmentVariables.TOKEN_EXCHANGE_VARS):
```
Comment (pvaneck) [inline sdk/identity/azure-identity/azure/identity/aio/_credentials/default.py:116]: Perhaps rename TOKEN_EXCHANGE_VARS to WORKLOAD_IDENTITY_VARS?

File: sdk/identity/azure-identity/azure/identity/_credentials/managed_identity.py [hunk 4]
```diff
@@ -64,17 +64,17 @@ def __init__(self, **kwargs: Any) -> None:
 
                 self._credential = CloudShellCredential(**kwargs)
         elif all(os.environ.get(var) for var in EnvironmentVariables.TOKEN_EXCHANGE_VARS):
```
Comment (pvaneck) [inline sdk/identity/azure-identity/azure/identity/_credentials/managed_identity.py:66]: Should we remove this check/implementation from `ManagedIdentityCredential` as outlined by @scottaddie  at the bottom of the issue described in https://github.com/Azure/azure-sdk-for-python/issues/26476?

Comment (xiangyan99) [reply sdk/identity/azure-identity/azure/identity/_credentials/managed_identity.py:66]: Removing it from MIC is a breaking change.

File: sdk/identity/azure-identity/azure/identity/aio/_credentials/default.py (hunk 1 above)
Comment (chlowell) [reply sdk/identity/azure-identity/azure/identity/aio/_credentials/default.py:117]: That's a good point! What do you think of adding `workload_identity_client_id` and preferring its value to `managed_identity_client_id`? It's ugly but avoids breaking applications and maintains the distinction between workload- and managed identity.

Comment (xiangyan99) [reply sdk/identity/azure-identity/azure/identity/aio/_credentials/default.py:117]: Let me ask this question:

Customer has the code:

```py
cred = DAC(managed_identity_client_id="xxx")
```

after they upgrade the identity library, do we want to keep exactly the same behavior?

Comment (xiangyan99) [reply sdk/identity/azure-identity/azure/identity/aio/_credentials/default.py]: I updated the code. How about this one? @chlowell

Comment (chlowell) [reply sdk/identity/azure-identity/azure/identity/aio/_credentials/default.py]: That looks correct to me: users setting `managed_identity_client_id` get the same behavior from the new credential, new users will (we hope) set `workload_identity_client_id` instead, and we prefer the new option when someone sets both.

File: sdk/identity/azure-identity/azure/identity/_credentials/default.py [hunk 5]
```diff
@@ -76,7 +80,7 @@ class DefaultAzureCredential(ChainedTokenCredential):
         Directory work or school accounts.
```
Comment (pvaneck) [inline sdk/identity/azure-identity/azure/identity/_credentials/default.py:80]: Add `:keyword  str workload_identity_client_id` somewhere in here And similarly in the aio version?

File: sdk/identity/azure-identity/azure/identity/_credentials/default.py [hunk 6]
```diff
@@ -64,6 +68,8 @@ class DefaultAzureCredential(ChainedTokenCredential):
         AZURE_TENANT_ID, if any. If unspecified, users will authenticate in their home tenants.
     :keyword str managed_identity_client_id: The client ID of a user-assigned managed identity. Defaults to the value
         of the environment variable AZURE_CLIENT_ID, if any. If not specified, a system-assigned identity will be used.
+    :keyword str workload_identity_client_id: The client ID of a user-assigned managed identity. Defaults to the value
+        of the environment variable AZURE_CLIENT_ID, if any. If not specified, a system-assigned identity will be used.
```
Comment (chlowell) [inline sdk/identity/azure-identity/azure/identity/_credentials/default.py:72]: Workload Identity supports service principals, and doesn't support managed identities yet:
```suggestion
    :keyword str workload_identity_client_id: The client ID of an identity assigned to the pod. Defaults to the value
        of the environment variable AZURE_CLIENT_ID, if any. If not specified, the pod's default identity will be used.
```

File: sdk/identity/azure-identity/azure/identity/aio/_credentials/default.py [hunk 7]
```diff
@@ -58,6 +62,8 @@ class DefaultAzureCredential(ChainedTokenCredential):
         **False**.
     :keyword str managed_identity_client_id: The client ID of a user-assigned managed identity. Defaults to the value
         of the environment variable AZURE_CLIENT_ID, if any. If not specified, a system-assigned identity will be used.
+    :keyword str workload_identity_client_id: The client ID of a user-assigned managed identity. Defaults to the value
+        of the environment variable AZURE_CLIENT_ID, if any. If not specified, a system-assigned identity will be used.
```
Comment (chlowell) [inline sdk/identity/azure-identity/azure/identity/aio/_credentials/default.py:66]:
```suggestion
    :keyword str workload_identity_client_id: The client ID of an identity assigned to the pod. Defaults to the value
        of the environment variable AZURE_CLIENT_ID, if any. If not specified, the pod's default identity will be used.
```

**Review by chlowell — APPROVED** [review]

**Comment by petergaultney** [comment]
did the SDK support Workload Identity in some other form before this PR was merged? I'm using 1.12 on an AKS cluster with Workload Identity enabled and it _seems_ to be working....  but now I see that the code for it apparently does not even exist in a released version?

**Comment by xiangyan99** [comment]
@petergaultney Yes. You are absolutely right! :)

**Pull request merged**